### PR TITLE
vhost_user_fs: Implement support for processing multiple requests in parallel

### PR DESCRIPTION
**Pull request by slp**
This patch series enables vhost_user_fs to process multiple requests in parallel by processing each one in an independent thread from a pool.

### CODE REVIEW

File: src/bin/vhost_user_fs.rs [hunk 1]
```diff
@@ -30,11 +31,13 @@ use virtio_bindings::bindings::virtio_net::*;
...
     VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC,
 };
-use vm_memory::GuestMemoryMmap;
+use vm_memory::{GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vm_virtio::queue::DescriptorChain;
 use vmm_sys_util::eventfd::EventFd;
 
 const QUEUE_SIZE: usize = 1024;
 const NUM_QUEUES: usize = 2;
+const THREAD_POOL_SIZE: usize = 64;
```
Comment (rbradford) [inline src/bin/vhost_user_fs.rs]: How did you choose this number?

File: vhost_user_fs/src/passthrough.rs [hunk 2]
```diff
@@ -48,7 +48,7 @@ struct InodeData {
 
 struct HandleData {
     inode: Inode,
-    file: Mutex<File>,
+    file: RwLock<File>,
```
Comment (rbradford) [inline vhost_user_fs/src/passthrough.rs]: So, we're actually looking at moving from RwLock to Mutex as it's got better performance.

File: src/bin/vhost_user_fs.rs (hunk 1 above)
Comment (slp) [reply src/bin/vhost_user_fs.rs]: That value comes from the usual queue depth of block devices. virtiofsd and QEMU's thread pool use the same number.

File: vhost_user_fs/src/passthrough.rs (hunk 2 above)
Comment (slp) [reply vhost_user_fs/src/passthrough.rs]: This situation is quite different than the one described in the paper referenced in #777. Here critical section is very large and costly (it even involves at least one syscall!), so the cost of lock management is insignificant in comparison.

In fact, I'd say the conclusions in that paper are only valid if your critical section is extremely small, as it's the case for the code in its benchmark, but that's another story.

Comment (rbradford) [reply vhost_user_fs/src/passthrough.rs]: @slp okay, please leave that comment on #777.

**Pull request merged**